File: Javascript/Arrays & Hashing/Check-If-One-String-Swap-Can-Make-Strings-Equal.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
const areAlmostEqual = (s1, s2) => {
  const s1Array = new Array(26).fill(0);
  const s2Array = new Array(26).fill(0);

  let counter = 0;

  for (let i = 0; i < s1.length; i++) {
    let charS1 = s1[i];
    let charS2 = s2[i];

    if (charS1 !== charS2) {
      counter++;
    }

    if (counter > 2) return false;

    s1Array[charS1.charCodeAt(0) - "a".charCodeAt(0)]++;
    s2Array[charS2.charCodeAt(0) - "a".charCodeAt(0)]++;
  }
 
  return s1Array.every((element, index) => element == s2Array[index]);
};

File: Rust/String/Check-If-One-String-Swap-Can-Make-Strings-Equal.rs
pub struct Solution;

impl Solution {
    pub fn are_almost_equal(s1: String, s2: String) -> bool {
        let vec: Vec<_> = s1
            .chars()
            .zip(s2.chars())
            .filter(|(c1, c2)| c1 != c2)
            .collect();

        match vec.len() {
            0 => true,
            2 => vec[0].0 == vec[1].1 && vec[0].1 == vec[1].0,
            _ => false,
        }

        // let mut s1_array: [u8; 26] = [0; 26];
        // let mut s2_array: [u8; 26] = [0; 26];

        // let mut counter = 0;

        // let s1 = s1.as_bytes();
        // let s2 = s2.as_bytes();

        // for index in 0..s1.len() {
        //     let char1 = s1[index];
        //     let char2 = s2[index];

        //     if char1 != char2 {
        //         counter += 1;
        //     }

        //     if counter > 2 {
        //         return false;
        //     }

        //     s1_array[(char1 - b'a') as usize] += 1;
        //     s2_array[(char2 - b'a') as usize] += 1;
        // }

        // s1_array
        //     .iter()
        //     .enumerate()
        //     .all(|(index, val)| *val == s2_array[index])
    }
}
fn main() {}
